src/mame/drivers/brother_lw30.cpp

```diff
@@ -0,0 +1,1144 @@
+			d = a;
+			e ^= input[i++];
+		}
+
+		return { c, d, e };
+	}
+
+	std::array<uint8_t, 416> gcr_encode_and_checksum(const uint8_t* input /* 256 bytes */) {
+		std::array<uint8_t, 416> output;
+		for(int i = 0; i < 51; i++)
+			gcr_encode_5_to_8(&input[i * 5], &output[i * 8]);
+
+		auto checksum = checksum_256_bytes(input);
+		std::array<uint8_t, 5> end_and_checksum{ input[255], checksum[0], checksum[1], checksum[2], 0x58 };
+		gcr_encode_5_to_8(&end_and_checksum[0], &output[408]);
+
+		return output;
+	}
+}
+
+image_init_result lw30_floppy_image_device::call_load()
+{
+	// create raw data from disk image
+	auto p = reinterpret_cast<uint8_t*>(ptr());
+	auto l = length();
+
+	if(l != 78 * 12 * 256)
+		return image_init_result::FAIL;
+
+	for(size_t track = 0; track < 78; track++) {
+		auto& data = tracks[track];
+		size_t i = 0;
+		for(size_t sector = 0; sector < 12; sector++) {
+			// according to check_track_and_sector
+			for(size_t j = 0; j < 5; j++)
+				data[i++] = 0xff;
+			data[i++] = 0xab;
+			data[i++] = sync_table[track] & 0xff;
+			data[i++] = sync_table[track] >> 8;
+			data[i++] = sync_table[sector] & 0xff;
+			data[i++] = sync_table[sector] >> 8;
+			data[i++] = 0xdd;
+			for(const auto& d : sector_header) // 16 bytes
+				data[i++] = d;
+			auto payload = gcr_encode_and_checksum(p + (track * 12 + sector) * 256); // 416 bytes
+			for(const auto& d : payload)
+				data[i++] = d;
+			for(const auto& d : sector_footer) // 11 bytes
+				data[i++] = d;
+		}
+	}
+
+	loaded = true;
+	return image_init_result::PASS;
+}
+
+void lw30_floppy_image_device::call_unload()
+{
+	if(dirty && !is_readonly()) {
+		//#todo
+		//if(reopen_for_write(filename()) == IMAGE_ERROR_SUCCESS)
+		//	fwrite(image.get(), image_length);
+	}
+
+	loaded = false;
+}
+
+#pragma endregion
+
+#pragma region(Sound)
+class brother_beep_device : public device_t, public device_sound_interface
+{
+public:
+	brother_beep_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock);
+
+protected:
+	// device-level overrides
+	void device_start() override;
+
+	// sound stream update overrides
+	void sound_stream_update(sound_stream &stream, std::vector<read_stream_view> const &inputs, std::vector<write_stream_view> &outputs) override;
+
+public:
+	DECLARE_WRITE_LINE_MEMBER(set_state);   // enable/disable sound output
+	void set_clock(uint32_t frequency);       // output frequency
+
+private:
+	sound_stream *m_stream;   /* stream number */
+	uint8_t m_state;          /* enable beep */
+	int m_frequency;          /* set frequency - this can be changed using the appropriate function */
+	int m_incr;               /* initial wave state */
+	int8_t m_signal;         /* current signal */
+};
+
+DEFINE_DEVICE_TYPE(BROTHER_BEEP, brother_beep_device, "brother_beep", "Brother Beep")
+
+static constexpr auto BROTHER_BEEP_RATE = 48000;
+
+brother_beep_device::brother_beep_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock)
```

Comment:
Are you sure you need a new device for a beeper?

Reply:
this beeper has a different restart behavior than the original one